lib/install/client/automatic_lifecycle.go

```diff
@@ -0,0 +1,119 @@
+/*
+Copyright 2018 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package client
+
+import (
+	"context"
+
+	installpb "github.com/gravitational/gravity/lib/install/proto"
+	"github.com/gravitational/gravity/lib/system/signals"
+
+	"github.com/gravitational/trace"
+)
+
+// HandleStatus handles the results of a completed operation.
+// It executes the handler corresponding to the outcome
+func (r *AutomaticLifecycle) HandleStatus(ctx context.Context, c *Client, status installpb.ProgressResponse_Status, err error) error {
+	switch {
+	case err == nil:
+		switch status {
+		case installpb.StatusUnknown:
+			if err1 := c.shutdown(ctx); err1 != nil {
+				c.WithError(err1).Warn("Failed to shut down.")
+			}
+			return nil
+		case installpb.StatusAborted:
+			return r.Abort(ctx, c)
+		}
+		// We received completion status
+		err = r.Complete(ctx, c, status)
+		return trace.Wrap(err)
+	case trace.IsEOF(err):
+		// Stream done but no completion event
+		if err1 := c.shutdown(ctx); err1 != nil {
+			c.WithError(err1).Warn("Failed to shut down.")
+		}
+		return nil
+	default:
+		if err1 := r.generateDebugReport(ctx, c); err1 != nil {
+			c.WithError(err1).Warn("Failed to generate debug report.")
+		}
+		if err1 := c.shutdown(ctx); err1 != nil {
```

Comment:
I decided to shutdown installer (and agents) on failure so the agent clients at least exit and users have the ability to use the same terminal and look around. The agents are configured to restart and reconnect, so it's a matter of `gravity resume` on the client to reattach to the progress loop.